### ISSUE TITLE
projection_layer should output logits value directly instead of log_softmax

### ISSUE THREAD

**Issue by liyang85105** [hkproj/pytorch-transformer#7]
The Projection Layer's output is torch.log_softmax(...)
```python
class ProjectionLayer(nn.Module):

    def __init__(self, d_model, vocab_size) -> None:
        super().__init__()
        self.proj = nn.Linear(d_model, vocab_size)

    def forward(self, x) -> None:
        # (batch, seq_len, d_model) --> (batch, seq_len, vocab_size)
        return torch.log_softmax(self.proj(x), dim = -1)
```
But if you choose to use the CrossEntropyLoss, it's already include the log and softmax ops inside, so this may cause a problem. Please correct me if I'm wrong.

**Comment by hkproj** [comment]
Hi @liyang85105 

You're right: the input to the `CrossEntropyLoss` should be **unnormalized logits**, which means the output of the linear layer **without** the `log_softmax`. The `CrossEntropyLoss` will normalize the logits using the `log_softmax` operation.

Quoting the CrossEntropyLoss documentation:

> The input is expected to contain the unnormalized logits for each class (which do not need to be positive or sum to 1, in general).

In our case, the input to the CrossEntropyLoss is already normalized by the `log_softmax`. Is it a problem?
Being the `softmax` and the `logarithm` both _monotonic functions_, the relative ordering of the logits is preserved by the `log_softmax` operation, that is, the most probable and the least probable token won't change by these two functions, but the relative distributions (or relative scale) of the elements may not be preserved, resulting in a slower training.

I'll remove the `log_softmax` and train again the model and let you know.

Thanks for the feedback and I hope you understood my explanation. In case you have more questions, feel free to comment.

你也可以用中文，我的汉语还可以。

**Comment by liyang85105** [comment]
Yes,  maybe some duplicate computations, but the result is also right.
你居然也会中文，棒棒哒~ 顺便说一句，非常棒的视频，干货满满，我从中也学到了很多东西，希望一直会有这么高质量的产出。

**Comment by hkproj** [comment]
谢谢你的反馈！我也希望能有时间做更多视频。我现在住在苏州 （我老婆是中国人），你来苏州的话给我发邮件！
